dpos/manager/dposmanager.go

```diff
@@ -340,14 +351,89 @@ func (d *DPOSManager) recoverAbnormalState(firstRecover bool) bool {
 		if firstRecover {
 			d.neededMajorityStatus = 1
 		} else {
-			d.neededMajorityStatus = len(arbiters)/3 + 1
+			d.neededMajorityStatus = len(arbiters)/2 + 1
 		}
+		d.waitRecover = make(chan struct{}, 1)
 		d.handler.RequestAbnormalRecovering()
+		go d.waitForRecover()
 		return true
 	}
 	return false
 }
 
+func (d *DPOSManager) waitForRecover() {
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+
+out:
+	for {
+		select {
+		case <-ticker.C:
+			break out
+
+		case <-d.waitRecover:
+			break out
+		}
+	}
+
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	close(d.waitRecover)
+	d.waitRecover = nil
+
+	if len(d.statusMap) == 0 {
+		log.Warn("wait for recover finished, but not received status")
+		return
+	}
+
+	var maxCount int
+	var maxCountMinViewOffset uint32
+	for k, v := range d.statusMap {
+		if maxCount < len(v) {
+			maxCount = len(v)
+			maxCountMinViewOffset = k
+		} else if maxCount == len(v) && maxCountMinViewOffset > k {
+			maxCountMinViewOffset = k
+		}
+	}
+	var status *dmsg.ConsensusStatus
+	startTimes := make([]int64, 0)
+	for _, v := range d.statusMap[maxCountMinViewOffset] {
+		if status == nil {
+			status = v
+		}
+		startTimes = append(startTimes, v.ViewStartTime.UnixNano())
+	}
+	sort.Slice(startTimes, func(i, j int) bool {
+		return startTimes[i] < startTimes[j]
+	})
+	medianTime := medianOf(startTimes)
+	sec := medianTime / 1e9
+	nsec := medianTime - sec
```

Comment:
Wrong math! should be nsec := medianTime - sec*1e9.

Reply:
fixed.....